geomesa-kafka/geomesa-kafka-09-datastore/src/main/scala/org/locationtech/geomesa/kafka09/ReplayKafkaConsumerFeatureSource.scala

```diff
@@ -0,0 +1,418 @@
+    val snapshot: Seq[GeoMessage] = messages.view
+      .drop(startIndex)
+      .takeWhile {
+        // stop at the first clear or when past the endTime
+        case c: Clear => false
+        case e => e.timestamp.getMillis >= endTime
+      }
+
+    if (snapshot.isEmpty) {
+      None
+    } else {
+      Some(ReplaySnapshotFeatureCache(replaySFT, time, snapshot))
+    }
+  }
+
+  private def readMessages(): Array[GeoMessage] = {
+    // use the liveSFT to decode becuase that's the type that was used to encode
+    val msgDecoder = new KafkaGeoMessageDecoder(liveSFT)
+
+    // start 1 ms earlier because there might be multiple messages with the same timestamp
+    val startTime = replayConfig.realStartTime.minus(1L)
+
+    val offsetRequest = FindOffset(msg => {
+      val key = msgDecoder.decodeKey(msg)
+      if (key.ts.isEqual(startTime)) 0 else if (key.ts.isAfter(startTime)) 1 else -1
+    })
+
+    // don't want to block waiting for more messages so specify  a timeout
+    val kafkaConsumer = kf.kafkaConsumer(topic, Map("consumer.timeout.ms" -> "250"))
+
+    try {
+      // required: there is only 1 partition;  validate??
+      val stream = kafkaConsumer.createMessageStreams(1, offsetRequest).head
+
+      val msgIds = scala.collection.mutable.HashSet.empty[(Int, Long)]
+
+      stream.iterator
+        .stopOnTimeout
+        .filter(m => msgIds.add((m.partition, m.offset))) // avoid replayed duplicates
+        .map(msgDecoder.decode)
+        .dropWhile(replayConfig.isBeforeRealStart)
+        .takeWhile(replayConfig.isNotAfterEnd)
+        .foldLeft(List.empty[GeoMessage])((seq, elem) => elem :: seq)
+        .toArray
+    } finally {
+      kafkaConsumer.shutdown()
+    }
+  }
+
+  override def close(): Unit = {}
+}
+
+object ReplayTimeHelper {
+
+  private[kafka09] val ff = CommonFactoryFinder.getFilterFactory2
+
+  val AttributeName: String = "KafkaLogTime"
+  val AttributeProp: PropertyName = ff.property(AttributeName)
+
+  def addReplayTimeAttribute(builder: SimpleFeatureTypeBuilder): Unit =
+    builder.add(AttributeName, classOf[Date])
+  
+  def toFilter(time: Instant): Filter =
+    ff.equals(AttributeProp, ff.literal(time.toDate))
+
+  def fromFilter(filter: PropertyIsEqualTo): Option[Long] = {
+    checkOrder(filter.getExpression1, filter.getExpression2)
+      .filter(pl => pl.name == AttributeName && pl.literal.getValue.isInstanceOf[Date])
+      .map(pl => new Instant(pl.literal.getValue.asInstanceOf[Date]).getMillis)
+  }
+}
+
+/** @param sft the [[SimpleFeatureType]] - must contain the replay time attribute
+  * @param replayTime the current replay time
+  */
+class ReplayTimeHelper(sft: SimpleFeatureType, replayTime: Long) {
+  import ReplayTimeHelper._
+
+  private val replayDate = new java.util.Date(replayTime)
+  private val builder = new SimpleFeatureBuilder(sft)
+  private val attrIndex = sft.indexOf(AttributeName)
+
+  require(attrIndex >= 0, s"Invalid SFT.  The $AttributeName attribute is missing.")
```

Comment:
At the risk of starting a flame war, Andrew is morally and typographically wrong.  Two spaces after a period, now and forever.